File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # create graph using adjacency list and indegree array
        adj = defaultdict(list)
        indegree = [0] * numCourses
        for course, pre in prerequisites:
            adj[pre].append(course)
            indegree[course] += 1                

        # bfs
        q = deque()
        for course in range(numCourses):
            if indegree[course] == 0:
                q.append(course)
        
        order = []
        while q:
            front = q.popleft()
            order.append(front)
            for course in adj[front]:
                indegree[course] -= 1
                if indegree[course] == 0:
                    q.append(course)
            

        # return true or false
        if len(order) < numCourses:
            return []
        return order

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] indegree = new int[numCourses];        
        for (int[] edge : prerequisites) {
            // [a,b] b->a
            graph.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
            indegree[edge[0]]++;
        }
        Queue<Integer> que = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) 
                que.offer(i);
        }
        List<Integer> res = new ArrayList<>();
        while (!que.isEmpty()) {
            int front = que.poll();
            res.add(front);
            if (!graph.containsKey(front)) continue;
            for (int nb : graph.get(front)) {
                indegree[nb]--;
                if (indegree[nb] == 0)
                    que.offer(nb);
            }
        }
        if (res.size() < numCourses) return new int[]{};
        int[] resArray = new int[numCourses];
        for (int i = 0; i < numCourses; i++)
            resArray[i] = res.get(i);
        return resArray;
        /**
            numCourses = 4, 
            [[1,0],[2,0],[3,1],[3,2]]
            0->1<->2
                 /^
                3 
            que = []     res = [0,3]
            indegree = [0 1 1 0]
                        0 1 2 3
         */
    }
}